assimilation_code/programs/preprocess/preprocess.f90

```diff
@@ -65,450 +120,453 @@ program preprocess
-   write(err_string, *) 'file ', trim(output_obs_kind_mod_file), &
-      ' exists and will not be overwritten: Please remove or rename'
-   call error_handler(E_ERR, 'preprocess', err_string, &
-      source, revision, revdate)
-endif
+call cannot_be_null(output_obs_def_mod_file, 'output_obs_def_mod_file')
+call cannot_be_null(output_obs_qty_mod_file, 'output_obs_qty_mod_file')
+
+call log_it('INPUT obs_def files:')
+
+do i = 1, MAX_OBS_TYPE_FILES
+   if(obs_type_files(i) == 'null') exit
+   call log_it(obs_type_files(i))
+   num_obs_type_files = i
+enddo
+
+call log_it('INPUT quantity files:')
+
+do i = 1, MAX_QUANTITY_FILES
+   if(quantity_files(i) == 'null') exit
+   call log_it(quantity_files(i))
+   num_quantity_files = i
+enddo
+
+! Open/Create the DEFAULT and OUTPUT files and set the unit numbers
+
+call open_file_for_read(input_obs_def_mod_file, 'input file ', obs_def_in_unit)
+call open_file_for_read(input_obs_qty_mod_file, 'input file ', obs_qty_in_unit)
+
+call open_file_for_write(output_obs_def_mod_file, 'output file ', overwrite_output, obs_def_out_unit)
+call open_file_for_write(output_obs_qty_mod_file, 'output file ', overwrite_output, obs_qty_out_unit)
 
 !______________________________________________________________________________
-! Preprocessing for the obs_kind module
-! Get all the type/kind strings from all of the obs_def files 
-! up front and then insert stuff.  Easier to error check and combine
-! duplicate kinds.
+! Preprocessing for the obs_qty module
+! Read in the quantity table and build index numbers for them. 
 
-! Initial number of types and kinds is 0
-num_types_found = 0
-num_kinds_found = 0
+! Copy over lines from obs_qty_template file up to the next insertion point
+linenum1 = 0
+call copy_until(obs_qty_in_unit,   input_obs_qty_mod_file, INSERT_INTS_STRING, linenum1, &
+                obs_qty_out_unit, output_obs_qty_mod_file, .false.)
 
-SEARCH_INPUT_FILES: do j = 1, num_input_files
-   if(file_exist(trim(input_files(j)))) then
-      ! Open the file for reading
-         in_unit = open_file(input_files(j), action='read')
-   else
-      ! If file does not exist it is an error
-      write(err_string, *) 'input_files ', trim(input_files(j)), &
-         ' does NOT exist (and must).'
-      call error_handler(E_ERR, 'preprocess', err_string, &
-         source, revision, revdate)
-   endif
+num_qtys_found = 0
 
-   ! Read until the ! BEGIN KIND LIST is found
-   linenum = 0
-   FIND_QTY_LIST: do
-
-      read(in_unit, 222, IOSTAT = ierr) line
-      ! If end of file, input file is incomplete or weird stuff happened
-      if(ierr /= 0) then
-         write(err_string, *) 'file ', trim(input_files(j)), &
-            ' does NOT contain ', kind_start_string
-         call error_handler(E_ERR, 'preprocess', err_string, &
-            source, revision, revdate)
-      endif
-      linenum = linenum + 1
+! QTY_STATE_VARIABLE is always defined (and always offset 0).
```

Comment:
here is where the 0 entry is forced by preprocess.  if there is nothing special about QTY_STATE_VARIABLE being 0, remove this.  (my suggestion is to keep it as-is for this pull request, and in a later update remove it if possible.  for now it doesn't hurt anything.  it might lead to confusing errors if someone tries to add additional name,value pairs to a quantity file for this quantity.  but i wouldn't hold up the pull request for this issue.